std/src/main/scala/algebra/std/map.scala

```diff
@@ -1,58 +1,132 @@
 package algebra
 package std
 
-import algebra.ring.{AdditiveMonoid, Rng}
+import algebra.ring._
+import algebra.std.util.StaticMethods.{addMap, initMutableMap, subtractMap, wrapMutableMap}
+import scala.collection.mutable
 
 package object map extends MapInstances
 
-trait MapInstances {
-  implicit def mapEq[K, V: Eq: AdditiveMonoid]: Eq[Map[K, V]] =
-    new MapEq[K, V]
-  
+trait MapInstances extends MapInstances3
+
+trait MapInstances3 extends MapInstances2 {
   implicit def mapRng[K, V: Rng] =
     new MapRng[K, V]
 }
 
-class MapEq[K, V](implicit ev: Eq[V], a: AdditiveMonoid[V]) extends Eq[Map[K, V]] {
+trait MapInstances2 extends MapInstances1 {
+  implicit def mapSemiring[K, V: Semiring]: MapSemiring[K, V] =
+    new MapSemiring[K, V]
+}
+
+trait MapInstances1 extends MapInstances0 {
+  implicit def mapGroup[K, V: Group]: MapGroup[K, V] =
+    new MapGroup[K, V]
+  implicit def mapAdditiveGroup[K, V: AdditiveGroup]: MapAdditiveGroup[K, V] =
+    new MapAdditiveGroup[K, V]
+}
+
+trait MapInstances0 {
+  implicit def mapEq[K, V: Eq: AdditiveMonoid]: Eq[Map[K, V]] =
```

Comment:
So -- this is a specific instance of a more general problem. There are two competing interpretations of maps:

1. sparse vector (missing keys are zeros)
2. dictionary (missing keys do not exist)

I think we had agreed that we were going to standardize on (1) and force users to make an explicit import/construction to get (2). Currently you'd have to build your own local instance of `MapEq` and/or hide the "default" one.

We could also provide neither by default. This is more annoying, and affects all users, but is a lot easier to handle (bring an implicit into scope, don't worry about ambiguity).